// postmark/dist/client/models/client/ClientOptions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ClientOptions;
(function (ClientOptions) {
    var Configuration = /** @class */ (function () {
        function Configuration(useHttps, requestHost, timeout) {
            this.useHttps = useHttps;
            this.requestHost = requestHost;
            this.timeout = timeout;
        }
        return Configuration;
    }());
    ClientOptions.Configuration = Configuration;
    var HttpMethod;
    (function (HttpMethod) {
        HttpMethod["GET"] = "GET";
        HttpMethod["POST"] = "POST";
        HttpMethod["DELETE"] = "DELETE";
        HttpMethod["PUT"] = "PUT";
        HttpMethod["OPTIONS"] = "OPTIONS";
        HttpMethod["HEAD"] = "HEAD";
        HttpMethod["PATCH"] = "PATCH";
    })(HttpMethod = ClientOptions.HttpMethod || (ClientOptions.HttpMethod = {}));
    var DefaultHeaderNames;
    (function (DefaultHeaderNames) {
        DefaultHeaderNames["SERVER_TOKEN"] = "X-Postmark-Server-Token";
        DefaultHeaderNames["ACCOUNT_TOKEN"] = "X-Postmark-Account-Token";
    })(DefaultHeaderNames = ClientOptions.DefaultHeaderNames || (ClientOptions.DefaultHeaderNames = {}));
})(ClientOptions = exports.ClientOptions || (exports.ClientOptions = {}));
